### PR TITLE
TOOL-369: minimal re-order and reproductional test case created

### PR DESCRIPTION
**Pull request by trapacska**
Originally the step was not able to parse versionCode if there was a comment at the end of the same line.

Now it can parse if:
- has comment at the end of the same line
- the value is not an integer but a variable
- syntax is kotlin
- any mixture of the above

Original imlementation covered:
Run the regex patterns on the lines: 
- `^versionCode (?P<version_code>.*)`: exact match `versionCode -anything-`
- `^versionName (?:"|')(?P<version_code>.*)(?:"|')` exact matches:
   - `versionName "-anything-"`
   - `versionName '-anything-"`
   - `versionName -<anything-'`
   - `versionName '-anything-'`

changed to:
- `^versionCode(?:\s|=)*(.*?)(?:\s|\/\/|$)`: exact match `versionCode -one or more whitespace or =- -anything- -until whitespace or end of line or //-`
- `^versionName(?:=|'|"|\s)+(.*?)(?:'|"|\s|\/\/|$)` exact matches:
   - `versionName -one or more whitespace or = or " or '- "- -anything- -until whitespace or end of line or // or ' or "-`
   - `versionName '-one or more whitespace or = or " or '- "- -anything- -until whitespace or end of line or // or ' or "-"`
   - `versionName -one or more whitespace or = or " or '- "- -<anything- -until whitespace or end of line or // or ' or "-'`
   - `versionName -one or more whitespace or = or " or '- "- '-anything- -until whitespace or end of line or // or ' or "-'`

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -92,75 +65,54 @@ func exportOutputs(outputs map[string]string) error {
...
-		if err != nil {
-			logFail("Failed to convert to string: %s, error: %s", configs.VersionCodeOffset, err)
-		}
-		newVersionCode += offset
+	var cfg config
+	if err := stepconf.Parse(&cfg); err != nil {
+		failf("Issue with input: %s", err)
 	}
+	stepconf.Print(cfg)
+	fmt.Println()
```
Comment (lpusok) [inline main.go]: Previously there was a check, which is now missing: errors.New("neither NewVersionCode nor NewVersionName are provided however, one of them is required")

File: main.go [hunk 2]
```diff
@@ -2,57 +2,30 @@ package main
...
+	// versionCode — A positive integer [...] -> https://developer.android.com/studio/publish/versioning
+	versionCodeRegexPattern = `^versionCode(?:\s|=)*(.*?)(?:\s|\/\/|$)`
+	// versionName — A string used as the version number shown to users [...] -> https://developer.android.com/studio/publish/versioning
+	versionNameRegexPattern = `^versionName(?:=|'|"|\s)+(.*?)(?:'|"|\s|\/\/|$)`
+)
 
-	return nil
+type config struct {
+	BuildGradlePth    string  `env:"build_gradle_path,file"`
+	NewVersionName    *string `env:"new_version_name"`
```
Comment (lpusok) [inline main.go]: Why not use just string and check if empty?

File: main.go [hunk 3]
```diff
@@ -92,75 +65,54 @@ func exportOutputs(outputs map[string]string) error {
...
-		}
-	}
-
-	if configs.VersionCodeOffset != "" {
-		offset, err := strconv.Atoi(configs.VersionCodeOffset)
-		if err != nil {
-			logFail("Failed to convert to string: %s, error: %s", configs.VersionCodeOffset, err)
-		}
-		newVersionCode += offset
+	var cfg config
```
Comment (lpusok) [inline main.go]: Change to Config or Configuration if the name is printed in the logs.

Comment (trapacska) [reply main.go]: the struct type name is printed, not the var name printed

File: main.go (hunk 2 above)
Comment (trapacska) [reply main.go]: we have this new feature in stepconf, and also it is a marshaler way of checking unset

Comment (lpusok) [reply main.go]: I am aware, just unsure if it is useful for string type as we discussed previously.

File: main.go [hunk 4]
```diff
@@ -78,6 +78,10 @@ func main() {
 	stepconf.Print(cfg)
 	fmt.Println()
 
+	if cfg.NewVersionName == nil && cfg.NewVersionCode == nil {
+		failf("Neither NewVersionCode nor NewVersionName are provided however, one of them is required.")
```
Comment (lpusok) [inline main.go]: Comma is in wrong place: "provided, however one of them is required" is better

**Pull request merged**